### PR TITLE
fix: allow users to connect walletconnect when all namespaces are optional

### PR DESCRIPTION
**Pull request by woodenfurniture**
## Description

Fixes issue where the app fails to connect a wallet with wallectconnect if all proposed chains are optional.

## Pull Request Type

- [x] :bug: Bug fix (Non-breaking Change: Fixes an issue)
- [ ] :hammer_and_wrench: Chore (Non-breaking Change: Doc updates, pkg upgrades, typos, etc..)
- [ ] :nail_care: New Feature (Breaking/Non-breaking Change)

## Issue (if applicable)

NA

## Risk

Low risk.

## Testing

Check app can connect to a dApp via walletconnect with only optional namespaces (see https://revoke.cash/ )

### CODE REVIEW

**Review by 0xApotheosis — CHANGES_REQUESTED** [review]
This causes some strange behaviour when connecting to dApps - I'm now getting two connected accounts in the dApp for every approved account in ShapeShift.

I suspect we are now telling the dApp we have both optional and required approvals for the requested and approved accounts.

File: src/context/WalletProvider/WalletProvider.tsx [hunk 1]
```diff
@@ -97,7 +96,7 @@ export interface InitialState {
   walletInfo: WalletInfo | null
   isConnected: boolean
   isDemoWallet: boolean
-  provider: MetaMaskLikeProvider | WalletConnectProvider | EthereumProviderType | null
+  provider: MetaMaskLikeProvider | EthereumProviderType | null
```
Comment (gomesalexandre) [inline src/context/WalletProvider/WalletProvider.tsx:99]: 💜

File: src/plugins/walletConnectToDapps/components/modals/SessionProposal.tsx [hunk 2]
```diff
@@ -47,27 +49,48 @@ const checkAllNamespacesHaveAccounts = (
...
+): SessionTypes.Namespaces => {
+  // tell lodash to concat array but merge everything else
+  const concatArrays = (objValue: unknown, srcValue: unknown) => {
+    if (Array.isArray(objValue)) {
+      return objValue.concat(srcValue)
+    }
+  }
+
+  // do a deep merge of the optional and required namespace approval objects
+  // but with a concat for the arrays so values stored on the same index aren't overwritten
```
Comment (gomesalexandre) [inline src/plugins/walletConnectToDapps/components/modals/SessionProposal.tsx:86]: 🧠

**Review by gomesalexandre — APPROVED** [review]
Tested against develop:

- Develop

Big sad app crash 😢 

- This diff

image

image

image

Note, Gnosis showing 0 accounts to connect is because Gnosis node is currently down, confirmed on the account page there are no accounts populated for it.

image

**Review by gomesalexandre — CHANGES_REQUESTED** [review]
@woodenfurniture I've been able to repro what @0xApotheosis has been seeing:

image
image

Requesting changes so we don't merge this guy just yet

**Review by gomesalexandre — APPROVED** [review]
gm

image
image

**Pull request merged**